src/coreclr/jit/gentree.cpp

```diff
@@ -2829,6 +2829,24 @@ bool GenTree::Compare(GenTree* op1, GenTree* op2, bool swapOK)
                 }
                 break;
 
+            case GT_CNS_LNG:
+            {
+                if (op1->AsLngCon()->gtLconVal == op2->AsLngCon()->gtLconVal)
+                {
+                    return true;
+                }
+                break;
+            }
+
+            case GT_CNS_DBL:
+            {
+                if (op1->AsDblCon()->isBitwiseEqual(op2->AsDblCon()))
```

Comment:
What’s the intent of GenTree::Compare here?

I feel like having an explanatory comment around why this type of comparison is fine for double is really necessary to elaborate on why NaN == NaN is okay for bitwise equivalent.

Such a comment would also help clarify if this could be updated to be more or less inclusive based on IEEE 754/ECMA 335 rules and whether or not this needs to ensure that float vs double typing for CNS_DBL needs to be considered since we only have the one node

Reply:
`GenTree::Compare` returns true if two trees are identical. This comment already exists on the function itself.

Reply:
There seems to be a lot of potentially dangerous nuance here.

Just looking at many of the callers, many (but not all) are having to explicitly skip handling of floating-point or NaNs/etc themselves when considering `GenTree::Compare`. It seems like one of those cases where it might be applicable to simply say that two floating-point nodes containing NaN are not equal ever and otherwise only if bitwise equal; just to help avoid potential problems.

Reply:
`gtFoldExprCompare` for example uses this to see "if two trees are identical, then folding is possible"

but of course it has to manually account for floating-point and skip it today, when in practice it could do it for everything except `NaN`

Reply:
`GenTree::Compare` is a syntactic equality check. It does not say anything about whether two trees compute to the same value under any form of comparison. This is a fundamental property of syntax vs semantics that callers need to take into account when they make use of it. For example, two syntactically identical calls for obvious reasons are not guaranteed to evaluate to the same value, even though `GenTree::Compare` returns true for them.

I do not see a good reason to pessimize `GenTree::Compare` for the callers that do use it correctly (like tail/head merging) to account for other callers potentially using it incorrectly. Those other callers should be fixed to realize that fundamentally `GenTree::Compare` cannot be used to guarantee in general that two trees produce the same value under any form of comparison.